### PR TITLE
Encode using media type instead of extension

### PR DESCRIPTION
**Pull request by deluxetom**
If an image had `image/x-webp` or another format supported by `Intervention` but not defined in `supportedFormats`, it would fail.

This uses `encodeByMediaType` instead which uses the proper values and available formats defined by `Intervention`

### CODE REVIEW

File: src/Api/Encoder.php [hunk 1]
```diff
@@ -101,35 +108,68 @@ public function run(ImageInterface $image): EncodedImageInterface
      */
     public function getFormat(ImageInterface $image): string
     {
-        $fm = (string) $this->getParam('fm');
```
Comment (ADmad) [inline src/Api/Encoder.php:104]: This needs to be preserved. If `fm` is specified then that's the format that should be returned irrespective of the actual media type.

File: src/Api/Encoder.php [hunk 2]
```diff
@@ -101,35 +108,68 @@ public function run(ImageInterface $image): EncodedImageInterface
...
-
-        /** @psalm-suppress RiskyTruthyFalsyComparison */
-        return array_search($image->origin()->mediaType(), static::supportedFormats(), true) ?: 'jpg';
     }
 
     /**
      * Get a list of supported image formats and MIME types.
      *
-     * @return array<string,string>
+     * @return array<string,MediaType>
```
Comment (ADmad) [inline src/Api/Encoder.php:124]: Changing the type can cause problems for users upgrading from 2.x. I prefer not breaking backwards compatibility unless absolutely necessary.

File: tests/Api/EncoderTest.php [hunk 3]
```diff
@@ -102,53 +115,43 @@ public function testRun(): void
...
-
-            if (function_exists('imagecreatefromwebp')) {
-                $this->assertMediaType($mock, 'image/webp')->once();
-            }
-
-            if (function_exists('imagecreatefromavif')) {
-                $this->assertMediaType($mock, 'image/avif')->once();
-            }
-        });
+        $this->assertSame('jpg', $this->encoder->setParams(['fm' => 'jpg'])->getFormat($this->getImageByMimeType('image/jpeg')));
```
Comment (ADmad) [inline tests/Api/EncoderTest.php:124]: We also need tests where `fm` does not match the image's mime type. `fm`'s value should take precedence in that case.

File: src/Api/Encoder.php (hunk 1 above)
Comment (deluxetom) [reply src/Api/Encoder.php:104]: it's preserved, its just moved there:
https://github.com/thephpleague/glide/pull/410/files#diff-63a111a0d299d1ea5fa389c1909b9577ed17e30e75c26627802b3cb3c8cdd606R89

File: tests/Api/EncoderTest.php (hunk 3 above)
Comment (deluxetom) [reply tests/Api/EncoderTest.php:124]: I added more tests to check the `fm` parameter always takes precedence

File: src/Api/Encoder.php (hunk 2 above)
Comment (deluxetom) [reply src/Api/Encoder.php]: fixed

File: src/Api/Encoder.php (hunk 1 above)
Comment (ADmad) [reply src/Api/Encoder.php:104]: Well directly calling this method won't take into account `fm` would it?  This is a public API.

Comment (ADmad) [reply src/Api/Encoder.php:104]: Sorry, I misread the code, `fm`'s value is accounted for.

File: src/Api/Api.php [hunk 4]
```diff
@@ -58,17 +58,11 @@ public function getImageManager(): ImageManager
     /**
      * Set the manipulators.
      *
-     * @param ManipulatorInterface[] $manipulators Collection of manipulators.
+     * @param array $manipulators Collection of manipulators.
      */
```
Comment (ADmad) [inline src/Api/Api.php:62]:
```suggestion
     * @throws \InvalidArgumentException
     */
```

File: src/Api/Api.php [hunk 5]
```diff
@@ -26,8 +26,8 @@ class Api implements ApiInterface
     /**
      * Create API instance.
      *
-     * @param ImageManager $imageManager Intervention image manager.
-     * @param array        $manipulators Collection of manipulators.
+     * @param ImageManager                    $imageManager Intervention image manager.
+     * @param list<ManipulatorInterface|null> $manipulators Collection of manipulators.
```
Comment (ADmad) [inline src/Api/Api.php:30]: I don't think `|null` is necessary

Comment (deluxetom) [reply src/Api/Api.php]: agreed, its been fixed

File: src/Api/Api.php (hunk 4 above)
Comment (deluxetom) [reply src/Api/Api.php:64]: @ADmad good catch,  I added the missing phpdoc

**Comment by ADmad** [comment]
Something went wrong, testsuite is failing.

**Comment by deluxetom** [comment]
> Something went wrong, testsuite is failing.

GitHub is having some issues, I can't restart the workflow, tests are all green locally:
```
> ./vendor/bin/phpunit
PHPUnit 11.5.1 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.4.1
Configuration: phpunit.xml.dist

...............................................................  63 / 216 ( 29%)
............................................................... 126 / 216 ( 58%)
............................................................... 189 / 216 ( 87%)
...........................                                     216 / 216 (100%)

Time: 00:00.327, Memory: 16.00 MB

OK (216 tests, 455 assertions)
```

I'll try to fix as soon as I can

**Comment by deluxetom** [comment]
@ADmad it seems `heic` isn't supported anymore, I'm not sure I understand why, it started when I merged the `3.x` branch from upstream.

**Comment by ADmad** [comment]
Aha, I remember now, `heic` decoding is broken on ubuntu-24.04. I forgot about that and updated the github actions config to use 24.04 instead of 22.04. Please revert the image for `tests` job to `ubuntu-22.04` and the error will be gone.

**Comment by deluxetom** [comment]
@ADmad all green now, let me know if there's anything you want me to update

**Comment by ADmad** [comment]
Looks good to go. I'll just have a look again with fresh eyes tomorrow before merging.

**Comment by ADmad** [comment]
@deluxetom I re-thought about the problem your PR aims to address and come up with an alternative which require much lesser code changes.

**Comment by ADmad** [comment]
Closing as the alternative PR is merged. Thank you @deluxetom.

**Pull request closed without merging**